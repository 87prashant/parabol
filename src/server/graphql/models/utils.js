import {GraphQLNonNull, GraphQLInputObjectType} from 'graphql';
import getRethink from 'server/database/rethinkDriver';
import mime from 'mime-types';
import {
  APP_CDN_USER_ASSET_SUBDIR,
  APP_MAX_AVATAR_FILE_SIZE
} from 'universal/utils/constants';
import protocolRelativeUrl from 'server/utils/protocolRelativeUrl';
import {s3SignPutObject} from 'server/utils/s3';
import path from 'path';

// Stringify an object to handle multiple errors
// Wrap it in a new Error type to avoid sending it twice via the originalError field
export const errorObj = obj => new Error(JSON.stringify(obj));

export const handleSchemaErrors = (errors) => {
  if (Object.keys(errors).length > 0) {
    throw errorObj(errors);
  }
};

// VERY important, otherwise eg a user could "create" a new team with an existing teamId & force join that team
export const ensureUniqueId = async (table, id) => {
  const r = getRethink();
  const res = await r.table(table).get(id);
  if (res) {
    throw errorObj({type: 'unique id collision'});
  }
};

export const ensureUserInOrg = async (userId, orgId) => {
  const r = getRethink();
  const inOrg = await r.table('User').get(userId)('orgs').contains(orgId);
  if (!inOrg) {
    throw errorObj({type: `user ${userId} does not belong to org ${orgId}`});
  }
  return true;
};

// if the add & update schemas have different required fields, use this
export const nonnullifyInputThunk = (name, inputThunk, requiredFieldNames) => {
  return new GraphQLInputObjectType({
    name,
    fields: () => {
      const newFields = inputThunk();
      requiredFieldNames.forEach(fieldName => {
        newFields[fieldName].type = new GraphQLNonNull(newFields[fieldName].type);
      });
      return newFields;
    }
  });
};

function getFields(context, astsParams = context.fieldNodes) {
  // for recursion...Fragments doesn't have many sets...
  const asts = Array.isArray(astsParams) ? astsParams : [astsParams];

  // get all selectionSets
  const selectionSets = asts.reduce((selections, source) => {
    selections.push(...source.selectionSet.selections);
    return selections;
  }, []);

  // return fields
  return selectionSets.reduce((list, ast) => {
    switch (ast.kind) {
      case 'Field' :
        // eslint-disable-next-line no-param-reassign
        list[ast.name.value] = true;
        return list;
      case 'InlineFragment':
        return {
          ...list,
          ...getFields(context, ast)
        };
      case 'FragmentSpread':
        return {
          ...list,
          ...getFields(context, context.fragments[ast.name.value])
        };
      default:
        throw new Error('Unsuported query selection');
    }
  }, {});
}

export function getRequestedFields(refs) {
  const fieldsObj = getFields(refs);
  return Object.keys(fieldsObj);
}

export function firstChange(possiblyUpdatedResult) {
  if (possiblyUpdatedResult.changes.length) {
    if (possiblyUpdatedResult.changes.length > 1) {
      console.warn('firstChange() detects more than 1 change, returning 1st.');
    }
    return possiblyUpdatedResult.changes[0];
  }
  return { new_val: undefined, old_val: undefined };
}

export function updatedOrOriginal(possiblyUpdatedResult, original) {
  /*
   * There will only be changes to return if there were changes made to the
   * DB. Therefore, we've got to check.
   */
  return firstChange(possiblyUpdatedResult).new_val || original;
}

export function getOldVal(possiblyUpdatedResult) {
  return firstChange(possiblyUpdatedResult).old_val;
}

export function getNewVal(resultWithReturnChanges) {
  return firstChange(resultWithReturnChanges).new_val;
}

export function makeEnumValues(constArr) {
  return constArr.reduce((values, constant) => {
    values[constant] = {value: constant};
    return values;
  }, {})
}

export function validateAvatarUpload(contentType, contentLength) {
  if (typeof process.env.CDN_BASE_URL === 'undefined') {
    throw errorObj({_error: 'CDN_BASE_URL environment variable is not defined'});
  }
  if (!contentType || !contentType.startsWith('image/')) {
    throw errorObj({_error: 'file must be an image'});
  }
  const ext = mime.extension(contentType);
  if (!ext) {
    throw errorObj({_error: `unable to determine extension for ${contentType}`});
  }
  if (contentLength > APP_MAX_AVATAR_FILE_SIZE) {
    throw errorObj({_error: 'avatar image is too large'});
  }
  return ext;
}

export function getS3PutUrl(contentType, contentLength, partialPath) {
  const parsedUrl = protocolRelativeUrl.parse(process.env.CDN_BASE_URL);
  const pathname = path.join(parsedUrl.pathname,
    APP_CDN_USER_ASSET_SUBDIR,
    partialPath
  );
  return s3SignPutObject(
    pathname,
    contentType,
    contentLength,
    'public-read'
  );
}
