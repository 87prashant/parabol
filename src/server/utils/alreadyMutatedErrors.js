import sendAuthRaven from 'server/utils/sendAuthRaven';

export const sendAlreadyJoinedIntegrationError = (authToken, integrationId) => {
  const breadcrumb = {
    message: 'You are already a part of this integration',
    category: 'Already mutated',
    data: {integrationId}
  };
  return sendAuthRaven(authToken, 'Easy there', breadcrumb);
};

export const sendAlreadyUpdatedIntegrationError = (authToken, integrationId) => {
  const breadcrumb = {
    message: 'Integration was already updated',
    category: 'Already mutated',
    data: {integrationId}
  };
  return sendAuthRaven(authToken, 'Easy there', breadcrumb);
};

export const sendAlreadyArchivedTeamError = (authToken, teamId) => {
  const breadcrumb = {
    message: 'Team was already archived',
    category: 'Team Already Archived',
    data: {teamId}
  };
  return sendAuthRaven(authToken, 'Easy there', breadcrumb);
}

export const sendAlreadyInactivatedUserError = (authToken, userId) => {
  const breadcrumb = {
    message: 'That user is already inactive. cannot inactivate twice',
    category: 'Already inactive',
    data: {userId}
  };
  return sendAuthRaven(authToken, 'Easy there', breadcrumb);
}
