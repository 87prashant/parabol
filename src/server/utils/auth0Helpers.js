import {ManagementClient} from 'auth0'
import getDotenv from '../../universal/utils/dotenv'

// Import .env and expand variables:
getDotenv()

export const clientId = process.env.AUTH0_CLIENT_ID
export const clientSecret = process.env.AUTH0_CLIENT_SECRET
export const domain = process.env.AUTH0_DOMAIN
export const managementClientId = process.env.AUTH0_MANAGEMENT_CLIENT_ID
export const managementClientSecret = process.env.AUTH0_MANAGEMENT_CLIENT_SECRET
export const managementDomain = process.env.AUTH0_MANAGEMENT_DOMAIN

export const auth0ManagementClient = new ManagementClient({
  clientId: managementClientId,
  clientSecret: managementClientSecret,
  domain: managementDomain
})
