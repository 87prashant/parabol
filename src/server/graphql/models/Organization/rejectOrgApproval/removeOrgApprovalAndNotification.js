import getRethink from 'server/database/rethinkDriver';
import {REQUEST_NEW_USER} from 'universal/utils/constants';

const removeOrgApprovalAndNotification = async (orgId, maybeEmails) => {
  const emails = Array.isArray(maybeEmails) ? maybeEmails : [maybeEmails];
  const r = getRethink();
  const {removedNotification} = await r({
    removedApproval: r.table('OrgApproval')
      .getAll(r.args(emails), {index: 'email'})
      .filter({orgId})
      .delete(),
    removedNotification: r.table('Notification')
      .getAll(orgId, {index: 'orgId'})
      .filter({
        type: REQUEST_NEW_USER
      })
      .filter((notification) => {
        return r.expr(emails).contains(notification('inviteeEmail'));
      })
      // get the inviterName
      .delete({returnChanges: true})('changes')
      .map((change) => change('old_val'))
      .default([])
  })
  return removedNotification;
};

export default removeOrgApprovalAndNotification;
