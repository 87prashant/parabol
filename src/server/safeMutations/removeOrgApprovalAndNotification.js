import getRethink from 'server/database/rethinkDriver';
import {REQUEST_NEW_USER} from 'universal/utils/constants';
import {APPROVED, DENIED, PENDING} from 'server/utils/serverConstants';

const removeOrgApprovalAndNotification = async (orgId, maybeEmails, type) => {
  const now = new Date();
  const {approvedBy, deniedBy} = type;
  const status = approvedBy ? APPROVED : DENIED;
  const emails = Array.isArray(maybeEmails) ? maybeEmails : [maybeEmails];
  const r = getRethink();
  const {removedNotifications} = await r({
    removedApproval: r.table('OrgApproval')
      .getAll(r.args(emails), {index: 'email'})
      .filter({orgId, status: PENDING})
      .update({
        status,
        approvedBy,
        deniedBy,
        updatedAt: now
      }),
    removedNotifications: r.table('Notification')
      .getAll(orgId, {index: 'orgId'})
      .filter({
        type: REQUEST_NEW_USER
      })
      .filter((notification) => {
        return r.expr(emails).contains(notification('inviteeEmail'));
      })
      // get the inviterName
      .delete({returnChanges: true})('changes')
      .map((change) => change('old_val'))
      .default([])
  });
  const notificationsToClear = {};
  removedNotifications.forEach((removedNotification) => {
    removedNotification.userIds.forEach((userId) => {
      notificationsToClear[userId] = notificationsToClear[userId] || [];
      notificationsToClear[userId].push(removedNotification.id);
    });
  });
  return notificationsToClear;
};

export default removeOrgApprovalAndNotification;
