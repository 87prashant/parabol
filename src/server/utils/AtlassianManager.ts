import AtlassianClientManager from 'universal/utils/AtlassianClientManager'
import makeAppLink from './makeAppLink'
import {ATLASSIAN_SCOPE} from 'universal/utils/constants'
import fetch from 'node-fetch'

class AtlassianManager extends AtlassianClientManager {
  static async init (code: string) {
    const queryParams = {
      grant_type: 'authorization_code',
      client_id: process.env.ATLASSIAN_CLIENT_ID,
      client_secret: process.env.ATLASSIAN_CLIENT_SECRET,
      code,
      redirect_uri: makeAppLink('auth/atlassian')
    }

    const uri = `https://auth.atlassian.com/oauth/token`

    const tokenRes = await fetch(uri, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(queryParams)
    })

    const tokenJson = await tokenRes.json()

    console.log('token', tokenJson)
    const {access_token: accessToken, refesh_token: refreshToken, error, scope} = tokenJson
    if (error) {
      throw new Error(`Atlassian: ${error}`)
    }
    const providedScope = scope.split(' ')
    const matchingScope =
      new Set([...ATLASSIAN_SCOPE.split(' '), ...providedScope]).size === providedScope.length
    if (!matchingScope) {
      throw new Error(`bad scope: ${scope}`)
    }
    return new AtlassianManager(accessToken, {fetch, refreshToken})
  }
}

export default AtlassianManager
