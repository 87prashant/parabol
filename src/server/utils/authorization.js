import {BILLING_LEADER} from 'universal/utils/constants';
import {qualifyingTiers, tierSupportsUpdateCheckInQuestion} from 'universal/utils/tierSupportsUpdateCheckInQuestion';
import getRethink from '../database/rethinkDriver';
import toTeamMemberId from 'universal/utils/relay/toTeamMemberId';

export const getUserId = (authToken) => {
  return authToken && typeof authToken === 'object' && authToken.sub;
};

export const isAuthenticated = (authToken) => Boolean(authToken);

export const isSuperUser = (authToken) => {
  const userId = getUserId(authToken);
  return userId && authToken.rol === 'su';
};

export const isOrgMember = (authToken, userOrgDoc, userId) => {

}
export const isTeamMember = (authToken, teamId) => {
  const {tms} = authToken;
  return Array.isArray(tms) && tms.includes(teamId);
};

export const isTeamLead = async (userId, teamId) => {
  const r = getRethink();
  const teamMemberId = toTeamMemberId(teamId, userId);
  return r.table('TeamMember').get(teamMemberId)('isLead').default(false).run();
};

export const getIsTeamLead = (teamMemberId) => {
  const r = getRethink();
  return r.table('TeamMember').get(teamMemberId)('isLead').default(false).run();
};

export const requireSU = (authToken) => {
  if (!isSuperUser(authToken)) {
    throw new Error('Unauthorized. Must be a super user to run this query.');
  }
};

// undefined orgId will disable the filter
export const getUserOrgDoc = (userId, orgId = '') => {
  const r = getRethink();
  return r.table('User').get(userId)('userOrgs')
    .filter({id: orgId})
    .nth(0)
    .default(null)
    .run();
};

export const isBillingLeader = (userOrgDoc) => {
  return (userOrgDoc && userOrgDoc.role === BILLING_LEADER);
};

export const isOrgLeaderOfUser = async (authToken, userId) => {
  const r = getRethink();
  return r.table('User')
    .get(authToken.sub)('userOrgs')
    .filter({
      role: BILLING_LEADER
    })('id')
    .do((leaderOrgs) => {
      return {
        leaderOrgs,
        memberOrgs: r.table('User')
          .get(userId)('userOrgs')('id')
      };
    })
    .do((res) => {
      return res('leaderOrgs')
        .union(res('memberOrgs'))
        .distinct()
        .count()
        .lt(res('leaderOrgs').count().add(res('memberOrgs').count()));
    });
};

export const requireTeamCanUpdateCheckInQuestion = async (teamId) => {
  const r = getRethink();

  const {tier} = await r
    .table('Team')
    .get(teamId)
    .pluck('tier');

  if (!tierSupportsUpdateCheckInQuestion(tier)) {
    throw new Error(
      `Unauthorized. Team billing tier must be one of {${qualifyingTiers.join(', ')}} to update the Check-in question. ` +
      `Actual tier is '${tier}'.`
    );
  }
  return true;
};
