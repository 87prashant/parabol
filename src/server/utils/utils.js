import mime from 'mime-types';
import {APP_MAX_AVATAR_FILE_SIZE} from '../../universal/utils/constants';

// todo put this in a legitify schema
export function validateAvatarUpload(contentType, contentLength) {
  if (typeof process.env.CDN_BASE_URL === 'undefined') {
    throw new Error('CDN_BASE_URL environment variable is not defined');
  }
  if (!contentType || !contentType.startsWith('image/')) {
    throw new Error('file must be an image');
  }
  const ext = mime.extension(contentType);
  if (!ext) {
    throw new Error(`unable to determine extension for ${contentType}`);
  }
  if (contentLength > APP_MAX_AVATAR_FILE_SIZE) {
    throw new Error('avatar image is too large');
  }
  return ext;
}
