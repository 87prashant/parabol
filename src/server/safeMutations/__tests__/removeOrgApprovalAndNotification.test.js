import DynamicSerializer from 'dynamic-serializer';
import MockDate from 'mockdate';
import MockDB from 'server/__tests__/setup/MockDB';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import newInvitee from 'server/__tests__/utils/newInvitee';
import getRethink from 'server/database/rethinkDriver';
import removeOrgApprovalAndNotification from 'server/safeMutations/removeOrgApprovalAndNotification';
import {BILLING_LEADER, REQUEST_NEW_USER} from 'universal/utils/constants';

MockDate.set(__now);
console.error = jest.fn();

describe('removeOrgApprovalAndNotification', () => {
  test('removes approvals and approval notifications, returns an object full of the notifications removed', async () => {
    // SETUP
    const r = getRethink();
    const mockDB = new MockDB();
    const dynamicSerializer = new DynamicSerializer();
    const invitee = newInvitee();
    await mockDB.init()
      .user(1, {userOrgs: [{role: BILLING_LEADER, id: mockDB.context.organization.id}]})
      .newOrgApproval({email: invitee.email})
      .newNotification(undefined, {type: REQUEST_NEW_USER, email: invitee.email});

    const orgId = mockDB.context.organization.id;
    // TEST
    const res = await removeOrgApprovalAndNotification(orgId, [invitee.email]);

    // VERIFY
    const db = await fetchAndSerialize({
      orgApproval: r.table('OrgApproval').getAll(invitee.email, {index: 'email'}).orderBy('email'),
      notification: r.table('Notification').getAll(orgId, {index: 'orgId'}).orderBy('userIds')
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();

    const [firstUser, secondUser] = mockDB.db.user;
    const notificationsToClear = [{deletedId: mockDB.context.notification.id}];
    expect(res[firstUser.id]).toEqual((notificationsToClear));
    expect(res[secondUser.id]).toEqual((notificationsToClear));
  });
});
