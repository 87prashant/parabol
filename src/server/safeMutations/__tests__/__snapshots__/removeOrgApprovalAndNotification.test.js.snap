// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeOrgApprovalAndNotification approves orgApproval, returns an object full of the notifications removed 1`] = `
Object {
  "notification": Array [],
  "orgApproval": Array [
    Object {
      "approvedBy": 1,
      "createdAt": 2017-01-01T07:00:00.000Z,
      "email": 2,
      "id": 0,
      "orgId": 3,
      "status": "APPROVED",
      "teamId": 4,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
}
`;

exports[`removeOrgApprovalAndNotification rejects approvals and approval notifications, returns an object full of the notifications removed 1`] = `
Object {
  "notification": Array [],
  "orgApproval": Array [
    Object {
      "createdAt": 2017-01-01T07:00:00.000Z,
      "deniedBy": 1,
      "email": 2,
      "id": 0,
      "orgId": 3,
      "status": "DENIED",
      "teamId": 4,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
}
`;
